### DISCUSSION TITLE
`router` in v3

### DISCUSSION THREAD

**Discussion by Ganbin** [chimurai/http-proxy-middleware#963] [Q&A]
Hi,

I have tried to search before to ask the question, but I didn't find what I was looking so let me ask.

I tried the HMP `v3.0.0-beta2`. I have a working solution with `v2.0.6` but it is not working with v3 and I don't understand why. I have tried multiple things but without success.

Here is my code:

```ts
import { createProxyMiddleware } from "http-proxy-middleware";
import { PluginOption } from "vite";

const proxyTable = {
  "localhost:3000": "http://localhost:3001",
  "testnet.localhost:3000": "https://testnet.mysite.come",
  "mainnet.localhost:3000": "https://mysite.come",
};
const proxy = createProxyMiddleware({
  router: proxyTable,
  changeOrigin: true,
}) as any;

export const proxyMiddleware: () => PluginOption = () => ({
  name: "custom-middleware",
  configureServer(server) {
    server.middlewares.use("/v1", proxy);
    server.middlewares.use("/v2", proxy);
    server.middlewares.use("/icons", proxy);
  },
});
```

I understand that something has changed with the `target` option where the path should be added in the `target` option but I'm not sure how this apply to the `router` option when we use such a mapping object.

Some of my try:

```diff
const proxyTable = {
-  "mainnet.localhost:3000": "https://mysite.come",
+  "mainnet.localhost:3000/v1": "https://mysite.come",
};

const proxyTable = {
-  "mainnet.localhost:3000": "https://mysite.come",
+  "mainnet.localhost:3000/v1": "https://mysite.come/v1",
};
-    server.middlewares.use("/v1", proxy);
+    server.middlewares.use(proxy);
```

Most of the time I get this error:
```shell
7:40:49 AM [vite] Internal server error: Must provide a proper URL as target
      at ProxyServer.<anonymous> (/Users/ganbin/Documents/app/node_modules/http-proxy/lib/http-proxy/index.js:69:35)
      at HttpProxyMiddleware.middleware (/Users/ganbin/Documents/app/node_modules/http-proxy-middleware/dist/http-proxy-middleware.js:23:32)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

But some of my tries just don't show any error but the routing is not working.

If I use `legacyCreateProxyMiddleware` it is working.

Any hint?

**Comment by TimJJTing** [comment]
Facing same problem here. It seems `target` must be present in option. But how should I specify `target` if all the rules are already specified in proxyTable?
